### PR TITLE
Git workflow script

### PR DESCRIPTION
**Pull request by lclarko**
- bash script utilizes getopts
 - Includes branch cloning 
 - Includes local branch cleanup

### CODE REVIEW

File: operations/workflow.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      read -p "Enter the branch label: " branch
+      read -p "Enter your GitHub Name (First name Last name): " username
+      read -p "Enter your email to associate commits to: " email
+      git clone --single-branch --branch $branch https://github.com/bcgov/GDX-Analytics-microservice.git $branch
+      cd $branch
+      git config user.name "${username}"
+      git config user.email "${email}"
+      ;;
+    c)
+      read -r -p "Enter branch name to delete: " rmbranch
```
Comment (mark-walle) [inline operations/workflow.sh:17]: @lclarko  can you check this dir exists before allowing to proceed? otherwise repeat this prompt?

Comment (mark-walle) [reply operations/workflow.sh:17]:
```
[microservice@ip-172-31-20-95 branch]$ rm fakedir/; echo $?
rm: cannot remove ‘fakedir/’: No such file or directory
1
[microservice@ip-172-31-20-95 branch]$
[microservice@ip-172-31-20-95 branch]$
rm: cannot remove ‘fakedir’: No such file or directory
[microservice@ip-172-31-20-95 branch]$ ./workflow.sh -c
Enter branch name to delete: fakedir
Are you sure you want to permanently delete branch folder /home/microservice/branch/fakedir? [y/N] y
Deleting /home/microservice/branch/fakedir ...
[microservice@ip-172-31-20-95 branch]$ echo $?
1
```

if it doesn't exist, either force the prompt again until user provides an existing folder, or exit (err code 1) with a similar error as rm gives before the rm -rf command runs later.

In the last command: `echo $?` I'm asking bash for the exit code of the last command. Your script gave no error message output, so I would have expected an exit code of 0. It actually exits with 1 (from the internal `rm -rf` itself returning 1 for trying to remove a `fakedir`, I think).

File: operations/workflow.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      elif [[ "$pipresponse" =~ ^([nN][oO]|[nN])$ ]]
+      then
+              printf "Please remove your pipenv before deleting branch folder\n"
+      else
+              printf "Exiting....\n"
+      fi
+      exit 1
+      ;;
+    \?)
+      echo "Invalid option: -$OPTARG try again..." >&2
```
Comment (mark-walle) [inline operations/workflow.sh:35]: list the options.

File: operations/workflow.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      ;;
+    :)
+      echo "Option -$OPTARG requires an argument." >&2
```
Comment (mark-walle) [inline operations/workflow.sh:38]: list the options.

File: operations/workflow.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+  
+BRANCH_PATH="$HOME/branch/"
+
+while getopts ":b :c" opt; do
+  case $opt in
+    b)
+      read -p "Enter the branch label: " branch
```
Comment (mark-walle) [inline operations/workflow.sh:8]: Can you check if the branch exists on the remote beforehand allowing the user to proceed?

with this command:

```
git ls-remote --exit-code --heads git@github.com:user/repo.git branch-name
```

exit code of 2 would tell you if there was no matching branch on the remote (halt or require input again).
exit code of 0 would mean it does exist (proceed).

**Review by mark-walle — CHANGES_REQUESTED** [review]

File: operations/workflow.sh (hunk 1 above)
Comment (lclarko) [reply operations/workflow.sh:17]: Added Check for DIR existing

File: operations/workflow.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      git ls-remote --heads ${REPO} ${branch} | grep ${branch} >/dev/null
+      if [ "$?" == "1" ] # Check exit code of git ls-remote
+      then
+              echo ""$branch" doesn't exist in "$REPO""
+              exit
+      fi
+      read -p "Enter your GitHub Name (First name Last name): " username
```
Comment (mark-walle) [inline operations/workflow.sh:25]: Can this just read "Enter your GitHub username"?

File: operations/workflow.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+# Usage: ./workflow.sh -opt                               #
+#                                                         #
+# Options: -b or -branch to clone a branch                #
+#          -c or -clean to delete a branch                #
+###########################################################
+
+BRANCH_PATH="$HOME/branch/"
+REPO="https://github.com/bcgov/GDX-Analytics-microservice.git"
+
+while getopts ":b(branch) :c(clean)" opt; do
```
Comment (mark-walle) [inline operations/workflow.sh:15]: Can there be a case where no flags are provided and as a result workflow.sh lists the available opts?

Right now the result is not very explanatory:
```
[microservice@ip-172-31-20-95 branch]$ ./workflow.sh 
[microservice@ip-172-31-20-95 branch]$ 
```

**Review by mark-walle — CHANGES_REQUESTED** [review]

File: operations/workflow.sh (hunk 5 above)
Comment (lclarko) [reply operations/workflow.sh:25]: I see that my variable name is confusing. Git config user.name is actually First and Last name, not username.
https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup/#_your_identity

Comment (lclarko) [reply operations/workflow.sh:25]: More direct - https://docs.github.com/en/github/getting-started-with-github/getting-started-with-git/setting-your-username-in-git

Comment (mark-walle) [reply operations/workflow.sh:25]: then it's not a "GitHub name"; let's just change it to "Enter your first and last name:" avoiding the confusion coming from multiple meanings of username.

Comment (lclarko) [reply operations/workflow.sh]: I've updated the prompt to be more straight forward

File: operations/workflow.sh (hunk 6 above)
Comment (lclarko) [reply operations/workflow.sh:15]: This should now be working. If no arg is passed, $OPTIND will equal 1, satisfying the condition and echoing the options

**Review by mark-walle — APPROVED** [review]
Looks good, thanks @lclarko

**Pull request merged**